In [1]:
# Imports and pip installations (if needed)
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
import numpy as np
from sklearn.model_selection import cross_val_score
from sklearn import linear_model
from sklearn.preprocessing import PolynomialFeatures
from sklearn.pipeline import Pipeline
from sklearn.datasets import load_iris
from sklearn.linear_model import LogisticRegression
from sklearn import svm
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier

# Part 1: Load the dataset

In [2]:
# Load the dataset (load remotely, not locally)
iris = load_iris()



In [3]:
iris

{'data': array([[5.1, 3.5, 1.4, 0.2],
        [4.9, 3. , 1.4, 0.2],
        [4.7, 3.2, 1.3, 0.2],
        [4.6, 3.1, 1.5, 0.2],
        [5. , 3.6, 1.4, 0.2],
        [5.4, 3.9, 1.7, 0.4],
        [4.6, 3.4, 1.4, 0.3],
        [5. , 3.4, 1.5, 0.2],
        [4.4, 2.9, 1.4, 0.2],
        [4.9, 3.1, 1.5, 0.1],
        [5.4, 3.7, 1.5, 0.2],
        [4.8, 3.4, 1.6, 0.2],
        [4.8, 3. , 1.4, 0.1],
        [4.3, 3. , 1.1, 0.1],
        [5.8, 4. , 1.2, 0.2],
        [5.7, 4.4, 1.5, 0.4],
        [5.4, 3.9, 1.3, 0.4],
        [5.1, 3.5, 1.4, 0.3],
        [5.7, 3.8, 1.7, 0.3],
        [5.1, 3.8, 1.5, 0.3],
        [5.4, 3.4, 1.7, 0.2],
        [5.1, 3.7, 1.5, 0.4],
        [4.6, 3.6, 1. , 0.2],
        [5.1, 3.3, 1.7, 0.5],
        [4.8, 3.4, 1.9, 0.2],
        [5. , 3. , 1.6, 0.2],
        [5. , 3.4, 1.6, 0.4],
        [5.2, 3.5, 1.5, 0.2],
        [5.2, 3.4, 1.4, 0.2],
        [4.7, 3.2, 1.6, 0.2],
        [4.8, 3.1, 1.6, 0.2],
        [5.4, 3.4, 1.5, 0.4],
        [5.2, 4.1, 1.5, 0.1],
  

In [4]:
type(iris)

sklearn.utils.Bunch

In [5]:
# Output the first 15 rows of the data
#need to convert this to df
iris_df = pd.DataFrame(data=iris.data, columns=iris.feature_names)
#add a new column based on target
iris_df['target'] = iris.target
iris_df.head(15)

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                 5.1               3.5                1.4               0.2   
1                 4.9               3.0                1.4               0.2   
2                 4.7               3.2                1.3               0.2   
3                 4.6               3.1                1.5               0.2   
4                 5.0               3.6                1.4               0.2   
5                 5.4               3.9                1.7               0.4   
6                 4.6               3.4                1.4               0.3   
7                 5.0               3.4                1.5               0.2   
8                 4.4               2.9                1.4               0.2   
9                 4.9               3.1                1.5               0.1   
10                5.4               3.7                1.5               0.2   
11                4.8               3.4                1.6               0.2   
12                4.8               3.0                1.4               0.1   
13                4.3               3.0                1.1               0.1   
14                5.8               4.0                1.2               0.2   

    target  
0        0  
1        0  
2        0  
3        0  
4        0  
5        0  
6        0  
7        0  
8        0  
9        0  
10       0  
11       0  
12       0  
13       0  
14       0

In [6]:

# Display a summary of the table information (number of datapoints, etc.)
iris_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
 4   target             150 non-null    int64  
dtypes: float64(4), int64(1)
memory usage: 6.0 KB


## About the dataset
Explain what the data is in your own words. What are your features and labels? What is the mapping of your labels to the actual classes?

Based on the documentation, this data has 4 features which include sepal length, sepal width, petal length, and petal width. Then we will also have 1 label, which are the species of the flowers. (There are 3 of them which includes, Iris-Setosa, Iris-Versicolour and Iris-Virginica)

For the labels for each class, there is a new added column called target(I added them, so it will be easier for me later on). It only has 3 values, 0,1,2. Each of them will be representing a different speicies of flower, in this case 'setosa', 'versicolor', 'virginica respectively.

# Part 2: Split the dataset into train and test

In [10]:
# Take the dataset and split it into our features (X) and label (y)
#all the features
x = iris_df[['sepal length (cm)', 'sepal width (cm)','petal length (cm)', 'petal width (cm)']]
y = iris_df['target']



In this case I will split the data in terms of x and y .


In [11]:
x

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                  5.1               3.5                1.4               0.2
1                  4.9               3.0                1.4               0.2
2                  4.7               3.2                1.3               0.2
3                  4.6               3.1                1.5               0.2
4                  5.0               3.6                1.4               0.2
..                 ...               ...                ...               ...
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

[150 rows x 4 columns]

In [12]:
y

0      0
1      0
2      0
3      0
4      0
      ..
145    2
146    2
147    2
148    2
149    2
Name: target, Length: 150, dtype: int64

In [13]:
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
x_train, x_test, y_train, y_test = train_test_split(x,y,test_size=0.1,train_size=0.9,random_state=42)

In [14]:
x_train

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
56                 6.3               3.3                4.7               1.6
104                6.5               3.0                5.8               2.2
69                 5.6               2.5                3.9               1.1
55                 5.7               2.8                4.5               1.3
132                6.4               2.8                5.6               2.2
..                 ...               ...                ...               ...
71                 6.1               2.8                4.0               1.3
106                4.9               2.5                4.5               1.7
14                 5.8               4.0                1.2               0.2
92                 5.8               2.6                4.0               1.2
102                7.1               3.0                5.9               2.1

[135 rows x 4 columns]

In this case, I split the data into trainning set and testing set. 90% of the data belong to trainning set and 10% are testing set.

# Part 3: Logistic Regression

In [15]:
# i. Use sklearn to train a LogisticRegression model on the training set
reg = LogisticRegression().fit(x_train,y_train)



In [16]:
first = x_test.head(1)

In [17]:
print(first)

    sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
73                6.1               2.8                4.7               1.2


In [18]:
# ii. For a sample datapoint, predict the probabilities for each possible class
prob = reg.predict_proba(first)
print(prob)

[[0.00345963 0.83721525 0.15932512]]


This means that this data has a probability of 0.3 % of being target 0, 83% of being  target 1 and 15% of being target 2.

In [24]:
y_test.head(1)

73    1
Name: target, dtype: int64

This is just so fun !  As we can see, based on the probabilities array, this flower has a 83% chance of being flower 1. And this dataset is belong to flower 1. This is very accurate.

In [25]:
# iii. Report on the score for Logistic regression model, what does the score measure?
scores = reg.score(x_test,y_test)
print(scores)

1.0


The best possible score we can have is 1.0. And we got a score of 1.0 meaning our model has correctly predicted all the test cases. 

Although this is a good thing, I feel a bit worried about a perfect predicted model because we only have 135 trainning sets, which in my opinion is pretty tiny sample size. For assignment 3, I think we have like 900 trainning sets, but it only has a score of about ~0.83.

This makes me wonder, does having a lot of trainning set really improve the accuracy? However, it might also mean this dataset has a well defined pattern.

Anyway, at least for this model, we have a perfect model ! Yay. 

In [26]:
# iv. Extract the coefficents and intercepts for the boundary line(s)
coef = reg.coef_
intercept = reg.intercept_

print("Coefficents : ", coef)
print("intercept: " , intercept)

Coefficents :  [[-0.4267278   0.97333912 -2.44447317 -1.03169392]
 [ 0.51330607 -0.22410519 -0.21526176 -0.85176821]
 [-0.08657827 -0.74923394  2.65973493  1.88346213]]
intercept:  [  9.49851186   1.91286418 -11.41137603]


# Part 4: Support Vector Machine

In [27]:
# i. Use sklearn to train a Support Vector Classifier on the training set
#https://scikit-learn.org/stable/modules/svm.html
clf = svm.SVC(probability=True).fit(x_train,y_train)

In [28]:

# ii. For a sample datapoint, predict the probabilities for each possible class

#I am still using the previous data to compare these two methods
print(x_test.head(1))
print(y_test.head(1))


    sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
73                6.1               2.8                4.7               1.2
73    1
Name: target, dtype: int64


In [29]:
clf_predict_prob = clf.predict_proba(x_test.head(1))
print(clf_predict_prob)

[[0.00686361 0.90258625 0.09055014]]


The probalitlity of being flower 0, 1, 2 is 0.6%, 89%, 9% respectively.

In [30]:
# iii. Report on the score for the SVM, what does the score measure?
clf_score = clf.score(x_test,y_test)
print(clf_score)

1.0


For this method, we also get a score of 1, meaning the model has predicted correclty for all the test cases.

# Part 5: Neural Network

In [31]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
neural = MLPClassifier(random_state=1, max_iter=2000).fit(x_train, y_train)



In [32]:
# ii. For a sample datapoint, predict the probabilities for each possible class
#print(x_test.head(1))
#print(y_test.head(1))
# I am still using the same data
neural_predict_proba = neural.predict_proba(x_test.head(1))
print(neural_predict_proba)


[[0.00122179 0.97072955 0.02804866]]


This is very good. This model says there is a 97% of this dataset of being flower 1. Which in fact is the type of flower that we are looking for.

In [33]:
# iii. Report on the score for the Neural Network, what does the score measure?

neural_score = neural.score(x_test,y_test)
print(neural_score)

1.0


This also gives us a score of 1. Meaning this model has correctly predicted all the test cases that we give to it.

In [34]:
# iv: Experiment with different options for the neural network, report on your best configuration (the highest score I was able to achieve was 0.8666)

neural2 = MLPClassifier(random_state=1, max_iter=300).fit(x_train, y_train)

/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/sklearn/neural_network/_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (300) reached and the optimization hasn't converged yet.
  warnings.warn(


In [35]:
neural_score2 = neural2.score(x_test,y_test)
print(neural_score2)

1.0


Hmm, the best score I got was a 1, which is having a iteration of 2000. I have that many iterations because when the iteration is small, it gives me a werid warning saying optimization hasn't converged yet. 

However, I tried to reduce the iteration to 300, which will cause an optimization not converged warning, but it still gives me a score of 1.0.  It is weird but who knows. Maybe neuron network is this powerful :) .

# Part 6: K-Nearest Neighbors

In [36]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier
neigh = KNeighborsClassifier(n_neighbors=3).fit(x_train,y_train)


In [37]:
# ii. For a sample datapoint, predict the probabilities for each possible class

neigh_predict_proba = neigh.predict_proba(x_test.head(1))
print(neigh_predict_proba)

[[0. 1. 0.]]


In [60]:
y_test.head(4)

73     1
18     0
118    2
78     1
Name: target, dtype: int64

In [38]:
neigh_predict_proba = neigh.predict_proba(x_test.head(4))
print(neigh_predict_proba)

[[0. 1. 0.]
 [1. 0. 0.]
 [0. 0. 1.]
 [0. 1. 0.]]


I can't believe what I just saw. This method is very confident about its predictions.  And they are all correct.  

In [62]:
# iii. Report on the score for kNN, what does the score measure?
neigh_score = neigh.score(x_test,y_test)
print(neigh_score)

1.0


The score I got is 1.0, which means all the test cases are being correctly predicted.

# Part 7: Conclusions and takeaways

In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?

For all the methods, I got a score of 1, meaning all of them correctly predicted all my test cases. This is good.

However, their performance of predicting probability varies: 

|methods| 0 | 1 | 2|
| --- | --- | --- | --- |
|Logistic Regression| 0.00345963 | 0.83721525 | 0.15932512 |
|Support Vector Machine| 0.00678268 | 0.89412469 | 0.09909262 |
|Neural Network|0.00122179 | 0.97072955 | 0.02804866 |
|K-Nearest Neighbors|0 | 1 | 0 |

Those probabilities are all generated by using my first test case.

In fact, I think for this dataset, all the models are performed equally well. Even though the probabilities of different methods are different, they are still predicting the correct result. Therefore, I don't think any method is performing better in this case. 

I think this happens because this dataset is having a good pattern, as a result , all the methods are able to predict the result.

However, if I have to choose which method perform the best, I would choose KNN because it has higher probability. I think this happens because KNN let its neighbors to determine its classification and this dataset is a well structured dataset meaning all the similar points are grouping together. As a result, KNN can guess the result easily.

One thing that I am surprise with is all the methods have a score of 1, meaning it can predict all the test cases correctly. I thought since we only have 150 datasets, the result might be a bit off, but in reality, everything is correctly predicted.